Add documentation for deliveries (#540)
livingdocsIO/documentation: content/guides/editor/publish-control/delivery/index.md

---
title: Publication Delivery
description: Trigger deliveries and see delivery build status.
weight: 6
---

For infinite products like an e-paper there is usually a “build” stage involved. With the new Deliveries feature we allow a user to trigger a remote system to kick off a delivery build and report the status of the build back to Livingdocs. This allows an editor to get a better view on the infinite product life cycle. These are the features:

![delivery](./delivery.png)

As soon as the document has been published, the delivery section will show up. Clicking on the build button will trigger the Webhook `document.build` which in turn will trigger the delivery build in the external system, as configured in `settings.js`.
The external build system can `POST` back the current build status with an optional message. This status and optional message is reflected in the delivery section of the publish control, as seen above.

Update from the external system back to livingdocs:

> `POST /api/v1/documents/360/addDeliveryStatus`

Reviewer: `addDeliveryStatus` with it's API, payload, stati should be documented in the public API doc -> https://docs.livingdocs.io/reference-docs/public-api/publications/. You can just look in the other examples and you will get the idea.

Reviewer: Here you can also add a comment to point to the publication API doc for more information.

Payload:

```json
{
  "reportId": "2SG2MAA9RwPn",
  "publicationId": 524,
  "deliveryHandle": "mobile",
  "status": "success",
  "message": "More info see <a href=\"https://google.ch\" target=\"_blank\">here</a>"
}
```

> ## Server Downstream Configuration

Reviewer: ```suggestion
## Add Webhooks and Deliveries to your Project Config
```

> Configure the webhook to trigger the external system to build a delivery.

Reviewer: ```suggestion
Add a Webhook to the Project Config to trigger the external system to build a delivery.
```

> `/example-server/projects/settings.js`

Reviewer: rm, in the doc we don't make references to the example-server

```js

> {

Reviewer: ```
// Project Config
{
  v: 2,
  ...
  webhooks: {
    ...
  }
}
```

Reviewer: I was thinking about how you can make it clear that webhooks is a toplevel property of the Project Config. I came to the conclusion that if you show the whole Project Config object with `v2` as an indicator, then it should be quite clear.

  ...
  webhooks: {
    active: true,
    configurations: [
      {
        handle: 'my-registered-build-endpoint',
        label: '...',
        description: '...',
        url: 'https://example.com/my-webhook-endpoint',
        secret: 'a-secret-token-to-sign-the-request',
        active: true,
        events: [
          {
            // the document.build event will be called when an editor presses the build button
            name: 'document.build',
            conditions: {
              // only call the Webhook when the delivery build called 'web' is triggered
              deliveryHandles: ['web']
            }
          }
        ]
      }
    ]
  },
  ...
}

```

> Configure the available deliveries for this project and optionally customize the build button labels.

Reviewer: ```suggestion
Configure the available deliveries for this project and optionally customize the build button labels in the Project Config.
```

> `/example-server/projects/<project-name>/index.js`

Reviewer: rm

```js

> {

Reviewer: ```
// Project Config
{
  v: 2,
  ...
  deliveries: [
    ...
  ]
}
```

  ...
  deliveries: [
    {
      handle: 'web',
      label: 'Website',
      isPrimary: true,
      icon: 'book-open',
      url: {
        origin: 'https://livingdocs.io',
        pathPattern: '/doc/:id'
      },
      build: {
        enabled: true,
        triggerButtonLabel: 'Build',
        retriggerButtonLabel: 'Build again',
        retryButtonLabel: 'Retry'
      }
    }
  ],
  ...
}
```

Configure which data-records have a build button and build status in the publish control.

> `/example-server/projects/<project-name>/content-types/<data-record-name>.js`

Reviewer: rm

> ```js

Reviewer: // Project Config contentTypes

{
  ...
  documentType: 'data-record',

>   publishControl: {},

Reviewer: i would remove that property, it has no relation in this case.

---
title: Publication Delivery
description: Trigger deliveries and see delivery build status.
weight: 6
---

> This guide will show you how trigger a delivery build and see its build status.

Reviewer: ```suggestion
This guide will show you how to trigger a delivery build and see its build status.
```
